File: backend/app/api/routes/videos.py
from uuid import UUID

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.schemas.video import (
    VideoListResponse,
    VideoResponse,
    VideoStatusResponse,
)
from app.services import video as video_service

router = APIRouter(prefix="/videos", tags=["videos"])


@router.post("", response_model=VideoResponse, status_code=201)
def upload_video(
    file: UploadFile = File(...),
    title: str = Form(...),
    recording_date: str = Form(...),
    participants: str = Form(""),
    context_notes: str = Form(""),
    db: Session = Depends(get_db),
):
    """Upload a video file with metadata."""
    from datetime import date as date_type

    from app.schemas.video import VideoCreate

    # Parse participants comma-separated string into list
    participants_list = (
        [p.strip() for p in participants.split(",") if p.strip()]
        if participants
        else None
    )

    try:
        parsed_date = date_type.fromisoformat(recording_date)
    except ValueError:
        raise HTTPException(status_code=422, detail="Invalid date format. Use YYYY-MM-DD.")

    video_data = VideoCreate(
        title=title,
        recording_date=parsed_date,
        participants=participants_list,
        context_notes=context_notes or None,
    )

    try:
        video = video_service.create_video(db, video_data, file)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    # Dispatch Celery processing task if available
    try:
        from app.tasks.video_processing import process_video

        process_video.delay(str(video.id))
    except (ImportError, Exception):
        pass  # Task not implemented yet

    return video


@router.get("", response_model=VideoListResponse)
def list_videos(
    skip: int = 0,
    limit: int = 20,
    db: Session = Depends(get_db),
):
    """List all videos with pagination."""
    videos, total = video_service.list_videos(db, skip=skip, limit=limit)
    return VideoListResponse(videos=videos, total=total)


@router.get("/{video_id}", response_model=VideoResponse)
def get_video(
    video_id: UUID,
    db: Session = Depends(get_db),
):
    """Get a single video by ID."""
    video = video_service.get_video(db, video_id)
    if video is None:
        raise HTTPException(status_code=404, detail="Video not found")
    return video


@router.get("/{video_id}/status", response_model=VideoStatusResponse)
def get_video_status(
    video_id: UUID,
    db: Session = Depends(get_db),
):
    """Get the processing status of a video (for polling)."""
    video = video_service.get_video(db, video_id)
    if video is None:
        raise HTTPException(status_code=404, detail="Video not found")
    return video

File: frontend/src/api/videos.ts
import apiClient from './client'
import type { Video } from '../types/video'

export interface VideoStatusResponse {
  id: string
  status: Video['status']
  error_message: string | null
}

export interface VideoListResponse {
  videos: Video[]
  total: number
}

export async function uploadVideo(
  file: File,
  metadata: {
    title: string
    recording_date: string
    participants?: string
    context_notes?: string
  },
  onProgress?: (pct: number) => void
): Promise<Video> {
  const form = new FormData()
  form.append('file', file)
  form.append('title', metadata.title)
  form.append('recording_date', metadata.recording_date)
  form.append('participants', metadata.participants ?? '')
  form.append('context_notes', metadata.context_notes ?? '')

  const { data } = await apiClient.post<Video>('/videos', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
    onUploadProgress(event: { loaded: number; total?: number }) {
      if (event.total && onProgress) {
        onProgress(Math.round((event.loaded * 100) / event.total))
      }
    },
  })
  return data
}

export async function getVideos(
  skip = 0,
  limit = 20
): Promise<VideoListResponse> {
  const { data } = await apiClient.get<VideoListResponse>('/videos', {
    params: { skip, limit },
  })
  return data
}

export async function getVideo(id: string): Promise<Video> {
  const { data } = await apiClient.get<Video>(`/videos/${id}`)
  return data
}

export async function getVideoStatus(
  id: string
): Promise<VideoStatusResponse> {
  const { data } = await apiClient.get<VideoStatusResponse>(
    `/videos/${id}/status`
  )
  return data
}
